client/finality-grandpa/src/environment.rs

```diff
@@ -1227,12 +1227,30 @@ where
 		},
 	};
 
-	if target_header.number() > best_header.number() {
-		return Err(Error::Safety(
-			"SelectChain returned a finality target higher than its best block".into(),
-		))
+	let is_descendent_of = is_descendent_of(&*client, None);
+
+	if target_header.number() > best_header.number() ||
+		target_header.number() == best_header.number() &&
```

Comment:
This entire check here is a little bit shit and that we then log some warning. This doesn't need to be an issue as between both of these calls there can be a reorg, leading to the case that the finality target isn't in the chain of the best block.

Reply:
@andresilva what about applying the strategy you initially suggested? That is don't have to separate calls as to modify the `finality_target()` method to return a tuple `(target, best)` in one shot?

Reply:
@bkchr I don't know what you mean, or rather I don't understand what I should do with your comment. This can happen due to a race condition but it can also happen that `SelectChain` is returning inconsistent results (which I suspect is the root cause of the issue on Westend).

@davxy I think that should still be the goal but wanted to get a quick fix PR to address the issue on Westend. After we change `SelectChain::finality_target` we might still want to be defensive against "bad" implementations and do this check anyway.

Reply:
I just wanted to say that there can be a race between getting the finality target and getting the best block. I mean it is totally legit that the best block for example goes backwards when the longest/best chain has a valid dispute. And in this case the finality target could also be greater than the "New best block"

Reply:
Indeed, and that is the result of doing two distinct async calls where the underlying blockchain state may change. We can avoid that by making sure that we get all the data in a single call to `SelectChain::finality_target` (what @davxy talks about in his comment). But we still don't control what is returned by `SelectChain`, and it probably still makes sense to have this check when we fix that and/or we need to make sure that the Polkadot `SelectChain` respects this invariant (e.g. after https://github.com/paritytech/substrate/pull/13289 we know that invariant holds for `LongestChain`).

Reply:
I can remove the logging (or downgrade to debug/trace) if you think it will cause confusion.

Reply:
I downgraded the logging to `debug`. I think it's useful to log this as it can help us catch "bad" `SelectChain` implementations, but it's useless for end users to be presented with this warning.